From 6d39fef27bf206d7993b4e4e7d3ad0c429a07330 Mon Sep 17 00:00:00 2001
From: Paul Adedeji <pauladedeji@google.com>
Date: Mon, 14 Jul 2025 12:05:34 -0700
Subject: [PATCH] [M138][ntp footer] Add menu closed callback for context menu.

Fixes a potential use-after-free issue on Mac Chromium builds, where
showing the context menu multiple times could lead to a dangling
pointer. This issue is not reproducible on Canary builds.

(cherry picked from commit 96113db753b093f6bfdf6e430b887a7125d6c89f)

Change-Id: I10ed3e531cd1d5943d9e27ca57a8c746c7718db9
Bug: 397946279, 431023939
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6680268
Commit-Queue: Paul Adedeji <pauladedeji@google.com>
Reviewed-by: Tibor Goldschwendt <tiborg@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#1480008}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6734125
Auto-Submit: Paul Adedeji <pauladedeji@google.com>
Cr-Commit-Position: refs/branch-heads/7204@{#2051}
Cr-Branched-From: d5de512dc9dc8ddfe4e6d71b0637578bb6158683-refs/heads/main@{#1465706}
---

--- a/chrome/browser/ui/views/new_tab_footer/footer_web_view.cc
+++ b/chrome/browser/ui/views/new_tab_footer/footer_web_view.cc
@@ -70,7 +70,9 @@
   context_menu_model_ = std::move(menu_model);
   context_menu_runner_ = std::make_unique<views::MenuRunner>(
       context_menu_model_.get(),
-      views::MenuRunner::HAS_MNEMONICS | views::MenuRunner::CONTEXT_MENU);
+      views::MenuRunner::HAS_MNEMONICS | views::MenuRunner::CONTEXT_MENU,
+      base::BindRepeating(&NewTabFooterWebView::HideCustomContextMenu,
+                          weak_factory_.GetWeakPtr()));
   context_menu_runner_->RunMenuAt(
       GetWidget(), nullptr, gfx::Rect(point, gfx::Size()),
       views::MenuAnchorPosition::kTopLeft, ui::mojom::MenuSourceType::kMouse,
@@ -78,9 +80,11 @@
 }
 
 void NewTabFooterWebView::HideCustomContextMenu() {
-  if (context_menu_runner_) {
+  if (context_menu_runner_ && context_menu_runner_->IsRunning()) {
     context_menu_runner_->Cancel();
   }
+  context_menu_runner_.reset();
+  context_menu_model_.reset();
 }
 
 bool NewTabFooterWebView::HandleKeyboardEvent(
